// web-audio-engine/lib/impl/AnalyserNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var util = require("../util");
var AudioNode = require("./AudioNode");
var AnalyserNodeDSP = require("./dsp/AnalyserNode");

var MaxFFTSize = 32768;
var MinFFTSize = 32;

var AnalyserNode = function (_AudioNode) {
  _inherits(AnalyserNode, _AudioNode);

  /**
   * @param {AudioContext} context
   */

  function AnalyserNode(context) {
    _classCallCheck(this, AnalyserNode);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(AnalyserNode).call(this, context, {
      inputs: [1],
      outputs: [1],
      channelCount: 1,
      channelCountMode: "max"
    }));

    _this._fftSize = 2048;
    _this._minDecibels = -100;
    _this._maxDecibels = -30;
    _this._smoothingTimeConstant = 0.8;
    return _this;
  }

  /**
   * @return {number}
   */


  _createClass(AnalyserNode, [{
    key: "getFftSize",
    value: function getFftSize() {
      return this._fftSize;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setFftSize",
    value: function setFftSize(value) {
      value = util.clip(value | 0, MinFFTSize, MaxFFTSize);
      value = util.toPowerOfTwo(value, Math.ceil);
      this._fftSize = value;
    }

    /**
     * @return {number}
     */

  }, {
    key: "getFrequencyBinCount",
    value: function getFrequencyBinCount() {
      return this._fftSize / 2;
    }

    /**
     * @return {number}
     */

  }, {
    key: "getMinDecibels",
    value: function getMinDecibels() {
      return this._minDecibels;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setMinDecibels",
    value: function setMinDecibels(value) {
      value = util.toNumber(value);
      /* istanbul ignore else */
      if (-Infinity < value && value < this._maxDecibels) {
        this._minDecibels = value;
      }
    }

    /**
     * @return {number}
     */

  }, {
    key: "getMaxDecibels",
    value: function getMaxDecibels() {
      return this._maxDecibels;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setMaxDecibels",
    value: function setMaxDecibels(value) {
      value = util.toNumber(value);
      /* istanbul ignore else */
      if (this._minDecibels < value && value < Infinity) {
        this._maxDecibels = value;
      }
    }

    /**
     * @return {number}
     */

  }, {
    key: "getSmoothingTimeConstant",
    value: function getSmoothingTimeConstant() {
      return this._smoothingTimeConstant;
    }

    /**
     * @param {number}
     */

  }, {
    key: "setSmoothingTimeConstant",
    value: function setSmoothingTimeConstant(value) {
      value = util.clip(util.toNumber(value), 0, 1);
      this._smoothingTimeConstant = value;
    }

    /**
     * @param {Float32Array} array
     */
    /* istanbul ignore next */

  }, {
    key: "getFloatFrequencyData",
    value: function getFloatFrequencyData() {
      throw new TypeError("NOT YET IMPLEMENTED");
    }

    /**
     * @param {Uint8Array} array
     */
    /* istanbul ignore next */

  }, {
    key: "getByteFrequencyData",
    value: function getByteFrequencyData() {
      throw new TypeError("NOT YET IMPLEMENTED");
    }

    /**
     * @param {Float32Array} array
     */
    /* istanbul ignore next */

  }, {
    key: "getFloatTimeDomainData",
    value: function getFloatTimeDomainData() {
      throw new TypeError("NOT YET IMPLEMENTED");
    }

    /**
     * @param {Uint8Array} array
     */
    /* istanbul ignore next */

  }, {
    key: "getByteTimeDomainData",
    value: function getByteTimeDomainData() {
      throw new TypeError("NOT YET IMPLEMENTED");
    }

    /**
     * @param {number} numberOfChannels
     */

  }, {
    key: "channelDidUpdate",
    value: function channelDidUpdate(numberOfChannels) {
      this.outputs[0].setNumberOfChannels(numberOfChannels);
    }
  }]);

  return AnalyserNode;
}(AudioNode);

Object.assign(AnalyserNode.prototype, AnalyserNodeDSP);

module.exports = AnalyserNode;
