// web-audio-engine/lib/impl/AudioSourceNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var AudioNode = require("./AudioNode");

/* istanbul ignore next */

var AudioSourceNode = function (_AudioNode) {
  _inherits(AudioSourceNode, _AudioNode);

  /**
   * @param {AudioContext} context
   */

  function AudioSourceNode(context) {
    _classCallCheck(this, AudioSourceNode);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(AudioSourceNode).call(this, context, {
      inputs: [],
      outputs: [1]
    }));
  }

  _createClass(AudioSourceNode, [{
    key: "getChannelCount",
    value: function getChannelCount() {
      return 0;
    }
  }, {
    key: "setChannelCount",
    value: function setChannelCount() {}
  }, {
    key: "getChannelCountMode",
    value: function getChannelCountMode() {
      return "explicit";
    }
  }, {
    key: "setChannelCountMode",
    value: function setChannelCountMode() {}
  }, {
    key: "getChannelInterpretation",
    value: function getChannelInterpretation() {
      return "discrete";
    }
  }, {
    key: "setChannelInterpretation",
    value: function setChannelInterpretation() {}
  }, {
    key: "enableOutputsIfNecessary",
    value: function enableOutputsIfNecessary() {}
  }, {
    key: "disableOutputsIfNecessary",
    value: function disableOutputsIfNecessary() {}
  }]);

  return AudioSourceNode;
}(AudioNode);

module.exports = AudioSourceNode;
