// web-audio-engine/lib/impl/AudioBufferSourceNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var util = require("../util");
var AudioScheduledSourceNode = require("./AudioScheduledSourceNode");
var AudioBuffer = require("./AudioBuffer");
var AudioBufferSourceNodeDSP = require("./dsp/AudioBufferSourceNode");

var AudioBufferSourceNode = function (_AudioScheduledSource) {
  _inherits(AudioBufferSourceNode, _AudioScheduledSource);

  /**
   * @param {AudioContext} context
   */

  function AudioBufferSourceNode(context) {
    _classCallCheck(this, AudioBufferSourceNode);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(AudioBufferSourceNode).call(this, context));

    _this._buffer = null;
    _this._audioData = null;
    _this._playbackRate = _this.addParam("control", 1);
    _this._detune = _this.addParam("control", 0);
    _this._loop = false;
    _this._loopStart = 0;
    _this._loopEnd = 0;
    _this._offset = 0;
    return _this;
  }

  /**
   * @return {AudioBuffer}
   */


  _createClass(AudioBufferSourceNode, [{
    key: "getBuffer",
    value: function getBuffer() {
      return this._buffer;
    }

    /**
     * @param {AudioBuffer} value
     */

  }, {
    key: "setBuffer",
    value: function setBuffer(value) {
      value = util.toImpl(value);

      /* istanbul ignore else */
      if (value instanceof AudioBuffer) {
        this._buffer = value;
        this._audioData = this._buffer.audioData;
        this.outputs[0].setNumberOfChannels(this._audioData.numberOfChannels);
      }
    }

    /**
     * @return {AudioParam}
     */

  }, {
    key: "getPlaybackRate",
    value: function getPlaybackRate() {
      return this._playbackRate;
    }

    /**
     * @return {AudioParam}
     */

  }, {
    key: "getDetune",
    value: function getDetune() {
      return this._detune;
    }

    /**
     * @return {boolean}
     */

  }, {
    key: "getLoop",
    value: function getLoop() {
      return this._loop;
    }

    /**
     * @param {boolean}
     */

  }, {
    key: "setLoop",
    value: function setLoop(value) {
      this._loop = !!value;
    }

    /**
     * @return {number}
     */

  }, {
    key: "getLoopStart",
    value: function getLoopStart() {
      return this._loopStart;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setLoopStart",
    value: function setLoopStart(value) {
      value = Math.max(0, util.toNumber(value));
      this._loopStart = value;
    }

    /**
     * @return {number}
     */

  }, {
    key: "getLoopEnd",
    value: function getLoopEnd() {
      return this._loopEnd;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setLoopEnd",
    value: function setLoopEnd(value) {
      value = Math.max(0, util.toNumber(value));
      this._loopEnd = value;
    }

    /**
     * @param {number} when
     * @param {number} offset
     * @param {number} duration
     */

  }, {
    key: "start",
    value: function start(when, offset, duration) {
      var _this2 = this;

      /* istanbul ignore next */
      if (this._startTime !== Infinity) {
        return;
      }

      duration = util.defaults(duration, Infinity);

      when = Math.max(this.context.currentTime, util.toNumber(when));
      offset = Math.max(0, offset | 0);
      duration = Math.max(0, util.toNumber(duration));

      this._startTime = when;
      this._startFrame = Math.round(when * this.sampleRate);
      this._offset = offset;

      if (duration !== Infinity) {
        this._stopFrame = Math.round((this._startTime + duration) * this.sampleRate);
      }

      this.context.sched(when, function () {
        _this2.outputs[0].enable();
      });

      this.dspStart();
    }
  }]);

  return AudioBufferSourceNode;
}(AudioScheduledSourceNode);

Object.assign(AudioBufferSourceNode.prototype, AudioBufferSourceNodeDSP);

module.exports = AudioBufferSourceNode;
