// web-audio-engine/lib/impl/BiquadFilterNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var AudioNode = require("./AudioNode");
var BiquadFilterNodeDSP = require("./dsp/BiquadFilterNode");

var BiquadFilterNode = function (_AudioNode) {
  _inherits(BiquadFilterNode, _AudioNode);

  /**
   * @param {AudioContext} context
   */

  function BiquadFilterNode(context) {
    _classCallCheck(this, BiquadFilterNode);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(BiquadFilterNode).call(this, context, {
      inputs: [1],
      outputs: [1],
      channelCount: 2,
      channelCountMode: "max"
    }));

    _this._type = BiquadFilterNodeDSP.LOWPASS;
    _this._frequency = _this.addParam("control", 350);
    _this._detune = _this.addParam("control", 0);
    _this._Q = _this.addParam("control", 1);
    _this._gain = _this.addParam("control", 0);

    _this.dspInit();
    _this.dspUpdateKernel(1);
    return _this;
  }

  /**
   * @return {string}
   */


  _createClass(BiquadFilterNode, [{
    key: "getType",
    value: function getType() {
      return this.toFilterTypeName(this._type);
    }

    /**
     * @param {string} value
     */

  }, {
    key: "setType",
    value: function setType(value) {
      value = this.fromFilterTypeName(value);
      /* istanbul ignore else */
      if (BiquadFilterNodeDSP.FilterTypes.indexOf(value) !== -1) {
        this._type = value;
      }
    }

    /**
     * @return {AudioParam}
     */

  }, {
    key: "getFrequency",
    value: function getFrequency() {
      return this._frequency;
    }

    /**
     * @return {AudioParam}
     */

  }, {
    key: "getDetune",
    value: function getDetune() {
      return this._detune;
    }

    /**
     * @return {AudioParam}
     */

  }, {
    key: "getQ",
    value: function getQ() {
      return this._Q;
    }

    /**
     * @return {AudioParam}
     */

  }, {
    key: "getGain",
    value: function getGain() {
      return this._gain;
    }

    /**
     * @param {Float32Array} frequencyHz
     * @param {Float32Array} magResponse
     * @param {Float32Array} phaseResponse
     */
    /* istanbul ignore next */

  }, {
    key: "getFrequencyResponse",
    value: function getFrequencyResponse() {
      throw new TypeError("NOT YET IMPLEMENTED");
    }

    /**
     * @param {number} numberOfChannels
     */

  }, {
    key: "channelDidUpdate",
    value: function channelDidUpdate(numberOfChannels) {
      this.dspUpdateKernel(numberOfChannels);
      this.outputs[0].setNumberOfChannels(numberOfChannels);
    }

    /**
     * @return {number}
     */

  }, {
    key: "getTailTime",
    value: function getTailTime() {
      return 0.2;
    }

    /**
     * @param {string} value
     * @return {number}
     */

  }, {
    key: "fromFilterTypeName",
    value: function fromFilterTypeName(value) {
      switch (value) {
        case "lowpass":
          return BiquadFilterNodeDSP.LOWPASS;
        case "highpass":
          return BiquadFilterNodeDSP.HIGHPASS;
        case "bandpass":
          return BiquadFilterNodeDSP.BANDPASS;
        case "lowshelf":
          return BiquadFilterNodeDSP.LOWSHELF;
        case "highshelf":
          return BiquadFilterNodeDSP.HIGHSHELF;
        case "peaking":
          return BiquadFilterNodeDSP.PEAKING;
        case "notch":
          return BiquadFilterNodeDSP.NOTCH;
        case "allpass":
          return BiquadFilterNodeDSP.ALLPASS;
      }
      return -1;
    }

    /**
     * @param {number} value
     * @return {string}
     */

  }, {
    key: "toFilterTypeName",
    value: function toFilterTypeName(value) {
      switch (value) {
        case BiquadFilterNodeDSP.LOWPASS:
          return "lowpass";
        case BiquadFilterNodeDSP.HIGHPASS:
          return "highpass";
        case BiquadFilterNodeDSP.BANDPASS:
          return "bandpass";
        case BiquadFilterNodeDSP.LOWSHELF:
          return "lowshelf";
        case BiquadFilterNodeDSP.HIGHSHELF:
          return "highshelf";
        case BiquadFilterNodeDSP.PEAKING:
          return "peaking";
        case BiquadFilterNodeDSP.NOTCH:
          return "notch";
        case BiquadFilterNodeDSP.ALLPASS:
          return "allpass";
      }
      /* istanbul ignore next */
    }
  }]);

  return BiquadFilterNode;
}(AudioNode);

Object.assign(BiquadFilterNode.prototype, BiquadFilterNodeDSP);

module.exports = BiquadFilterNode;
