// web-audio-engine/lib/impl/ConvolverNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var util = require("../util");
var AudioNode = require("./AudioNode");
var AudioBuffer = require("./AudioBuffer");
var ConvolverNodeDSP = require("./dsp/ConvolverNode");

var ConvolverNode = function (_AudioNode) {
  _inherits(ConvolverNode, _AudioNode);

  /**
   * @param {AudioContext} context
   */

  function ConvolverNode(context) {
    _classCallCheck(this, ConvolverNode);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(ConvolverNode).call(this, context, {
      inputs: [1],
      outputs: [1],
      channelCount: 2,
      channelCountMode: "clamped-max"
    }));

    _this._buffer = null;
    _this._audioData = null;
    _this._normalize = true;
    return _this;
  }

  /**
   * @param {number} value
   */


  _createClass(ConvolverNode, [{
    key: "setChannelCount",
    value: function setChannelCount(value) {
      value = util.clip(value | 0, 1, 2);
      _get(Object.getPrototypeOf(ConvolverNode.prototype), "setChannelCount", this).call(this, value);
    }

    /**
     * @param {string} value
     */

  }, {
    key: "setChannelCountMode",
    value: function setChannelCountMode(value) {
      /* istanbul ignore else */
      if (value === "clamped-max" || value === "explicit") {
        _get(Object.getPrototypeOf(ConvolverNode.prototype), "setChannelCountMode", this).call(this, value);
      }
    }

    /**
     * @return {AudioBuffer}
     */

  }, {
    key: "getBuffer",
    value: function getBuffer() {
      return this._buffer;
    }

    /**
     * @param {AudioBuffer} value
     */

  }, {
    key: "setBuffer",
    value: function setBuffer(value) {
      value = util.toImpl(value);

      /* istanbul ignore else */
      if (value instanceof AudioBuffer) {
        this._buffer = value;
        this._audioData = this._buffer.audioData;
      }
    }

    /**
     * @return {boolean}
     */

  }, {
    key: "getNormalize",
    value: function getNormalize() {
      return this._normalize;
    }

    /**
     * @param {boolean} value
     */

  }, {
    key: "setNormalize",
    value: function setNormalize(value) {
      this._normalize = !!value;
    }

    /**
     * @param {number} numberOfChannels
     */

  }, {
    key: "channelDidUpdate",
    value: function channelDidUpdate(numberOfChannels) {
      numberOfChannels = Math.min(numberOfChannels, 2);

      this.outputs[0].setNumberOfChannels(numberOfChannels);
    }
  }]);

  return ConvolverNode;
}(AudioNode);

Object.assign(ConvolverNode.prototype, ConvolverNodeDSP);

module.exports = ConvolverNode;
