// web-audio-engine/lib/impl/GainNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var AudioNode = require("./AudioNode");
var GainNodeDSP = require("./dsp/GainNode");

var GainNode = function (_AudioNode) {
  _inherits(GainNode, _AudioNode);

  /**
   * @param {AudioContext} context
   */

  function GainNode(context) {
    _classCallCheck(this, GainNode);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(GainNode).call(this, context, {
      inputs: [1],
      outputs: [1],
      channelCount: 2,
      channelCountMode: "max"
    }));

    _this._gain = _this.addParam("audio", 1);
    return _this;
  }

  /**
   * @return {AudioParam}
   */


  _createClass(GainNode, [{
    key: "getGain",
    value: function getGain() {
      return this._gain;
    }

    /**
     * @param {number} numberOfChannels
     */

  }, {
    key: "channelDidUpdate",
    value: function channelDidUpdate(numberOfChannels) {
      this.outputs[0].setNumberOfChannels(numberOfChannels);
    }
  }]);

  return GainNode;
}(AudioNode);

Object.assign(GainNode.prototype, GainNodeDSP);

module.exports = GainNode;
