// web-audio-engine/lib/impl/IIRFilterNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var util = require("../util");
var AudioNode = require("./AudioNode");
var IIRFilterNodeDSP = require("./dsp/IIRFilterNode");

var IIRFilterNode = function (_AudioNode) {
  _inherits(IIRFilterNode, _AudioNode);

  /**
   * @param {AudioContext} context
   * @param {object}       opts
   * @param {Float32Array} opts.feedforward
   * @param {Float32Array} opts.feedback
   */

  function IIRFilterNode(context, opts) {
    _classCallCheck(this, IIRFilterNode);

    opts = opts || /* istanbul ignore next */{};

    var feedforward = util.defaults(opts.feedforward, [0]);
    var feedback = util.defaults(opts.feedback, [1]);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(IIRFilterNode).call(this, context, {
      inputs: [1],
      outputs: [1],
      channelCount: 2,
      channelCountMode: "max"
    }));

    _this._feedforward = feedforward;
    _this._feedback = feedback;

    _this.dspInit();
    _this.dspUpdateKernel(1);
    return _this;
  }

  /**
   * @param {Float32Array} frequencyHz
   * @param {Float32Array} magResponse
   * @param {Float32Array} phaseResponse
   */
  /* istanbul ignore next */


  _createClass(IIRFilterNode, [{
    key: "getFrequencyResponse",
    value: function getFrequencyResponse() {
      throw new TypeError("NOT YET IMPLEMENTED");
    }

    /**
     * @return {Float32Array}
     */

  }, {
    key: "getFeedforward",
    value: function getFeedforward() {
      return this._feedforward;
    }

    /**
     * @return {Float32Array}
     */

  }, {
    key: "getFeedback",
    value: function getFeedback() {
      return this._feedback;
    }

    /**
     * @param {number} numberOfChannels
     */

  }, {
    key: "channelDidUpdate",
    value: function channelDidUpdate(numberOfChannels) {
      this.dspUpdateKernel(numberOfChannels);
      this.outputs[0].setNumberOfChannels(numberOfChannels);
    }
  }]);

  return IIRFilterNode;
}(AudioNode);

Object.assign(IIRFilterNode.prototype, IIRFilterNodeDSP);

module.exports = IIRFilterNode;
