// web-audio-engine/lib/impl/OscillatorNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var util = require("../util");
var AudioScheduledSourceNode = require("./AudioScheduledSourceNode");
var PeriodicWave = require("./PeriodicWave");
var OscillatorNodeDSP = require("./dsp/OscillatorNode");

var OscillatorTypes = PeriodicWave.BasicWaveForms;
var DefaultPeriodicWaves = {};

var OscillatorNode = function (_AudioScheduledSource) {
  _inherits(OscillatorNode, _AudioScheduledSource);

  /**
   * @param {AudioContext} context
   */

  function OscillatorNode(context) {
    _classCallCheck(this, OscillatorNode);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(OscillatorNode).call(this, context));

    _this._frequency = _this.addParam("audio", 440);
    _this._detune = _this.addParam("audio", 0);
    _this._type = "sine";
    _this._periodicWave = _this.buildPeriodicWave(_this._type);
    _this._waveTable = null;
    _this.dspInit();
    return _this;
  }

  /**
   * @return {string}
   */


  _createClass(OscillatorNode, [{
    key: "getType",
    value: function getType() {
      return this._type;
    }

    /**
     * @param {string} value
     */

  }, {
    key: "setType",
    value: function setType(value) {
      /* istanbul ignore else */
      if (OscillatorTypes.indexOf(value) !== -1) {
        this._type = value;
        this._periodicWave = this.buildPeriodicWave(value);
        this._waveTable = this._periodicWave.getWaveTable();
      }
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getFrequency",
    value: function getFrequency() {
      return this._frequency;
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getDetune",
    value: function getDetune() {
      return this._detune;
    }

    /**
     * @param {PeriodicWave} periodicWave
     */

  }, {
    key: "setPeriodicWave",
    value: function setPeriodicWave(periodicWave) {
      periodicWave = util.toImpl(periodicWave);

      /* istanbul ignore else */
      if (periodicWave instanceof PeriodicWave) {
        this._type = "custom";
        this._periodicWave = periodicWave;
        this._waveTable = this._periodicWave.getWaveTable();
      }
    }

    /**
     * @return {PeriodicWave}
     */

  }, {
    key: "getPeriodicWave",
    value: function getPeriodicWave() {
      return this._periodicWave;
    }

    /**
     * @param {string} type
     * @return {PeriodicWave}
     */

  }, {
    key: "buildPeriodicWave",
    value: function buildPeriodicWave(type) {
      var sampleRate = this.context.sampleRate;
      var key = type + ":" + sampleRate;

      /* istanbul ignore else */
      if (!DefaultPeriodicWaves[key]) {
        var periodicWave = new PeriodicWave({ sampleRate: sampleRate }, { constraints: false });

        periodicWave.generateBasicWaveform(type);

        DefaultPeriodicWaves[key] = periodicWave;
      }

      return DefaultPeriodicWaves[key];
    }
  }]);

  return OscillatorNode;
}(AudioScheduledSourceNode);

Object.assign(OscillatorNode.prototype, OscillatorNodeDSP);

module.exports = OscillatorNode;
