// web-audio-engine/lib/impl/SpatialPannerNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var BasePannerNode = require("./BasePannerNode");
var SpatialPannerNodeDSP = require("./dsp/SpatialPannerNode");

var SpatialPannerNode = function (_BasePannerNode) {
  _inherits(SpatialPannerNode, _BasePannerNode);

  /**
   * @param {AudioContext}
   */

  function SpatialPannerNode(context) {
    _classCallCheck(this, SpatialPannerNode);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(SpatialPannerNode).call(this, context));

    _this._positionX = _this.addParam("audio", 0);
    _this._positionY = _this.addParam("audio", 0);
    _this._positionZ = _this.addParam("audio", 0);
    _this._orientationX = _this.addParam("audio", 0);
    _this._orientationY = _this.addParam("audio", 0);
    _this._orientationZ = _this.addParam("audio", 0);
    return _this;
  }

  /**
   * @param {AudioParam}
   */


  _createClass(SpatialPannerNode, [{
    key: "getPositionX",
    value: function getPositionX() {
      return this._positionX;
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getPositionY",
    value: function getPositionY() {
      return this._positionY;
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getPositionZ",
    value: function getPositionZ() {
      return this._positionZ;
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getOrientationX",
    value: function getOrientationX() {
      return this._positionX;
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getOrientationY",
    value: function getOrientationY() {
      return this._positionY;
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getOrientationZ",
    value: function getOrientationZ() {
      return this._positionZ;
    }
  }]);

  return SpatialPannerNode;
}(BasePannerNode);

Object.assign(SpatialPannerNode.prototype, SpatialPannerNodeDSP);

module.exports = SpatialPannerNode;
